### PR TITLE
Add an option to override the global `suspend_timeout` setting

### PR DESCRIPTION
**Pull request by mrtmm**
Introduce a new XBlock attribute for overriding the `suspend_timeout`
setting.

### CODE REVIEW

File: hastexo/jobs.py [hunk 1]
```diff
@@ -69,7 +66,9 @@ def run(self):
             stacks = Stack.objects.select_for_update().filter(
                 suspend_timestamp__isnull=False
             ).filter(
-                suspend_timestamp__lt=cutoff
+                suspend_by__isnull=False
+            ).filter(
+                suspend_by__lt=timezone.now()
```
Comment (fghaas) [inline hastexo/jobs.py:71]: Hmmm maybe I'm missing something, but we don't backfill the `suspend_by` field, correct? 

So, if 
* we're not backfilling but 
* we are looking only for objects where `suspend_by` is not null, and where its content is a date in the past, 
 
doesn't that mean we effectively *don't* suspend the labs that have ceased sending keepalives less than `suspend_timeout` ago, i.e. that were last touched in the two minutes prior to migration, until the learner opens them again?

Comment (mrtmm) [reply hastexo/jobs.py:86]: Good point, I updated the PR.

File: hastexo/jobs.py [hunk 2]
```diff
@@ -66,15 +66,32 @@ def run(self):
 
         # Get stacks to suspend
         with transaction.atomic():
-            stacks = Stack.objects.select_for_update().filter(
+            stacks_1 = Stack.objects.select_for_update().filter(
```
Comment (fghaas) [inline hastexo/jobs.py:69]: To make this easier to understand and more readable, I'd suggest to call this variable `stacks_without_suspend_by`.

File: hastexo/jobs.py [hunk 3]
```diff
@@ -66,15 +66,32 @@ def run(self):
...
             ).filter(
                 suspend_timestamp__lt=cutoff
             ).filter(
                 status__in=states
             ).exclude(
                 provider__exact=''
-            ).order_by('suspend_timestamp')[:concurrency]
+            )
+
+            stacks_2 = Stack.objects.select_for_update().filter(
```
Comment (fghaas) [inline hastexo/jobs.py:81]: ... and this, `stacks_with_suspend_by`.

File: hastexo/jobs.py [hunk 4]
```diff
@@ -66,15 +66,32 @@ def run(self):
...
+            ).filter(
+                suspend_by__lt=timezone.now()
+            ).filter(
+                status__in=states
+            ).exclude(
+                provider__exact=''
+            )
+
+            stacks = stacks_1.union(stacks_2).order_by(
+                'suspend_timestamp')[:concurrency]
```
Comment (fghaas) [inline hastexo/jobs.py:94]: I'm not quite sure what `[:concurrency]` does here, could you fill me in please?

**Review by fghaas — COMMENTED** [review]
This looks good to me now, modulo one style (naming) note and a question.

Just checking to see if I understand this correctly: if a learner comes to one lab where the course author has set `suspend_timeout` to, say, 1800, and then they move on to another lab where the XBlock has no `suspend_timeout` attribute, then as they move away from the lab the default 120s timeout applies. Correct?

File: hastexo/jobs.py (hunk 2 above)
Comment (mrtmm) [reply hastexo/jobs.py:69]: Ack.

File: hastexo/jobs.py (hunk 3 above)
Comment (mrtmm) [reply hastexo/jobs.py:81]: Ack.

File: hastexo/jobs.py (hunk 4 above)
Comment (mrtmm) [reply hastexo/jobs.py:94]: It limits the number of stacks to suspend in one job run.

**Review by mrtmm — COMMENTED** [review]
>  if a learner comes to one lab where the course author has set suspend_timeout to, say, 1800, and then they move on to another lab where the XBlock has no suspend_timeout attribute, then as they move away from the lab the default 120s timeout applies. Correct?
Yes.

File: hastexo/jobs.py (hunk 4 above)
Comment (fghaas) [reply hastexo/jobs.py]: Ah! 💡 

Thanks!

File: hastexo/jobs.py [hunk 5]
```diff
@@ -66,15 +66,33 @@ def run(self):
 
         # Get stacks to suspend
         with transaction.atomic():
-            stacks = Stack.objects.select_for_update().filter(
+            stacks_without_suspend_by = Stack.objects.select_for_update().filter(
```
Comment (fghaas) [inline hastexo/jobs.py:69]: Whoops, now `flake8` chokes on this line. :) Might want to add `# noqa: E501` at the end rather than break this line.

**Review by fghaas — APPROVED** [review]
This looks good to me now. If the tests pass I'll go ahead and merge this.

**Pull request merged**